File: scripts/make-icons.js
/* Genera los iconos PNG de marca de La Positiva sin dependencias externas.
   El motivo es un plato visto desde arriba: crema con borde dorado sobre
   terracota. Se usa en vez de la foto de un plato, que no sirve como icono.
   Uso:  node scripts/make-icons.js                                        */
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

const TERRACOTA = [0xb4, 0x47, 0x2a];
const DORADO    = [0xc8, 0x96, 0x3e];
const CREMA     = [0xf3, 0xec, 0xe2];

function crc32(buf) {
  let c, crc = 0xffffffff;
  for (let n = 0; n < buf.length; n++) {
    c = (crc ^ buf[n]) & 0xff;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    crc = c ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length, 0);
  const body = Buffer.concat([Buffer.from(type, 'ascii'), data]);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(body), 0);
  return Buffer.concat([len, body, crc]);
}

function png(size, pixel) {
  // Cada scanline lleva un byte de filtro (0 = None) y luego RGB.
  const raw = Buffer.alloc(size * (1 + size * 3));
  let p = 0;
  for (let y = 0; y < size; y++) {
    raw[p++] = 0;
    for (let x = 0; x < size; x++) {
      const c = pixel(x, y);
      raw[p++] = c[0]; raw[p++] = c[1]; raw[p++] = c[2];
    }
  }
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);
  ihdr.writeUInt32BE(size, 4);
  ihdr[8] = 8;    // 8 bits por canal
  ihdr[9] = 2;    // color type 2 = RGB
  return Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
    chunk('IHDR', ihdr),
    chunk('IDAT', zlib.deflateSync(raw, { level: 9 })),
    chunk('IEND', Buffer.alloc(0))
  ]);
}

// Antialiasing simple por supermuestreo de 3x3.
function makePlate(size, plateRatio) {
  const c = (size - 1) / 2;
  const rOut = size * plateRatio;
  const rIn = rOut * 0.72;
  return (x, y) => {
    let acc = [0, 0, 0];
    for (let sy = 0; sy < 3; sy++) {
      for (let sx = 0; sx < 3; sx++) {
        const dx = x + (sx + 0.5) / 3 - 0.5 - c;
        const dy = y + (sy + 0.5) / 3 - 0.5 - c;
        const d = Math.sqrt(dx * dx + dy * dy);
        const col = d <= rIn ? CREMA : (d <= rOut ? DORADO : TERRACOTA);
        acc[0] += col[0]; acc[1] += col[1]; acc[2] += col[2];
      }
    }
    return [Math.round(acc[0] / 9), Math.round(acc[1] / 9), Math.round(acc[2] / 9)];
  };
}

const outDir = path.join(__dirname, '..', 'icons');
fs.mkdirSync(outDir, { recursive: true });

const jobs = [
  // Los normales usan un plato grande; el maskable lo achica para
  // respetar la zona segura del 80% que recortan Android y Chrome.
  ['icon-192.png', 192, 0.34],
  ['icon-512.png', 512, 0.34],
  ['icon-512-maskable.png', 512, 0.26],
  ['apple-touch-icon.png', 180, 0.34]
];

for (const [name, size, ratio] of jobs) {
  const buf = png(size, makePlate(size, ratio));
  fs.writeFileSync(path.join(outDir, name), buf);
  console.log(name.padEnd(24), size + 'x' + size, buf.length + ' bytes');
}

File: scripts/make-icons.py
# -*- coding: utf-8 -*-
"""Genera los iconos de La Positiva a partir del logo de marca.

Reemplaza al viejo make-icons.js, que DIBUJABA un plato por codigo porque
todavia no habia logo. Ahora que existe la marca, los iconos salen de ella.

  Fuente:  icons/origen/logo-la-positiva.jpg
  Uso:     python scripts/make-icons.py

Necesita Pillow (pip install pillow). No corre en el build de Vercel: los PNG
quedan versionados en el repo. Se corre a mano solo si cambia el logo.
"""
import os
import sys

try:
    from PIL import Image
except ImportError:
    sys.exit("Falta Pillow. Instalalo con:  pip install pillow")

AQUI = os.path.dirname(os.path.abspath(__file__))
RAIZ = os.path.dirname(AQUI)
ORIGEN = os.path.join(RAIZ, 'icons', 'origen', 'logo-la-positiva.jpg')
DESTINO = os.path.join(RAIZ, 'icons')


def color_de_fondo(im):
    """El bordo de la marca, sacado del borde de la imagen.

    Se usa la mediana y no el promedio: el JPEG ensucia los pixeles con ruido
    de compresion y un solo pixel raro correria el promedio. Ademas sirve para
    rellenar el margen del icono maskable con el color exacto, en vez de
    estirar pixeles con artefactos.
    """
    px = im.load()
    an, al = im.size
    borde = []
    for x in range(an):
        borde.append(px[x, 0])
        borde.append(px[x, al - 1])
    for y in range(al):
        borde.append(px[0, y])
        borde.append(px[an - 1, y])

    canales = []
    for i in range(3):
        vals = sorted(c[i] for c in borde)
        canales.append(vals[len(vals) // 2])
    return tuple(canales)


def cuadrado(im):
    """Recorta al centro por si el original no viniera cuadrado."""
    an, al = im.size
    if an == al:
        return im
    lado = min(an, al)
    izq = (an - lado) // 2
    arr = (al - lado) // 2
    return im.crop((izq, arr, izq + lado, arr + lado))


def generar(im, fondo, size, escala):
    """Un icono de `size` px con el logo ocupando `escala` del ancho.

    escala < 1 deja margen alrededor: es lo que necesita el icono maskable,
    porque Android y Chrome recortan hasta un 20% del borde y si el logo va a
    sangre te comen las puntas de la P.
    """
    lienzo = Image.new('RGB', (size, size), fondo)
    lado = max(1, int(round(size * escala)))
    # LANCZOS es el mejor remuestreo de Pillow para ampliar y para reducir.
    logo = im.resize((lado, lado), Image.LANCZOS)
    off = (size - lado) // 2
    lienzo.paste(logo, (off, off))
    return lienzo


def main():
    if not os.path.exists(ORIGEN):
        sys.exit("No encuentro el logo en: %s" % ORIGEN)

    im = cuadrado(Image.open(ORIGEN).convert('RGB'))
    fondo = color_de_fondo(im)
    print("origen: %dx%d  |  color de marca: #%02x%02x%02x" % (im.size + fondo))

    if im.size[0] < 512:
        print("AVISO: el original mide %dpx. El icono de 512 sale ampliado y "
              "se va a ver menos nitido. Si aparece un archivo mas grande, "
              "reemplazalo y volve a correr esto." % im.size[0])

    trabajos = [
        # nombre,                    tamanio, escala del logo
        ('icon-192.png',                 192, 1.00),
        ('icon-512.png',                 512, 1.00),
        ('icon-512-maskable.png',        512, 0.78),   # zona segura de Android
        ('apple-touch-icon.png',         180, 1.00),
    ]

    os.makedirs(DESTINO, exist_ok=True)
    for nombre, size, escala in trabajos:
        salida = os.path.join(DESTINO, nombre)
        icono = generar(im, fondo, size, escala)
        # Paleta de 256 colores: la imagen es bordo y blanco, asi que a simple
        # vista es identica, pero el PNG pesa como un 30% menos. La textura
        # ampliada del JPEG es lo que lo inflaba.
        icono = icono.convert('P', palette=Image.ADAPTIVE, colors=256)
        icono.save(salida, 'PNG', optimize=True)
        print("%-26s %dx%d  %d bytes" % (nombre, size, size, os.path.getsize(salida)))


if __name__ == '__main__':
    main()
